<!--
Copyright (c) 2014, Goncalo Cabrita, ISR University of Coimbra
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:
    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
    * Neither the name of Clearpath Robotics nor the
      names of its contributors may be used to endorse or promote products
      derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-->

<launch>

    <arg name="gui" default="true"/>
    
    <!-- Open the Gazebo world -->
    <include file="$(find hratc2015_framework)/launch/world.launch">
      <arg name="gui" value="$(arg gui)" />
    </include>
    <!-- Spawn the Husky robot -->
    <include file="$(find hratc2015_framework)/launch/husky.launch"/>
    
    
    <!-- Load the controllers -->
    <rosparam file="$(find hratc2015_framework)/controllers/controllers.yaml" command="load"/>

    <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false"
      output="screen" args="joint_state_controller arm_controller ptu_d46_controller"/>

    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"
      respawn="false" output="screen">
      <!--<remap from="/joint_states" to="/husky/joint_states" />-->
    </node>
   
   
    <node pkg="hratc2015_framework" type="gps_to_odom" name="gps_to_odom" />

    <!-- EKF for absolute localization (clearpath) -->
    <node pkg="robot_pose_ekf_cp" type="robot_pose_ekf_cp" name="robot_pose_ekf" output="screen">
        <!--<remap from="odom" to="encoder"/> -->
        <remap from="imu_data" to="imu/data"/>
        <remap from="gps" to="gps/odom"/>
        <param name="output_frame" value="odom"/>
        <param name="freq" value="10.0"/>
        <param name="sensor_timeout" value="1.0"/>
        <param name="odom_used" value="false"/>
        <param name="imu_used" value="true"/>
        <param name="imu_absolute" value="true"/>
        <param name="vo_used" value="false"/>
        <param name="gps_used" value="true"/>
        <param name="debug" value="false"/>
        <param name="self_diagnose" value="false"/>
    </node>
    
    <!-- The minefield frame -->
    <node name="minefield_static_tf_publisher" pkg="hratc2015_framework" type="minefield_static_tf_publisher" output="screen">
      <rosparam file="$(find hratc2015_framework)/settings/minefield.yaml" command="load" />
    </node>

    <!-- The coil signal simulator - required when not using the real robot -->
    <node name="coilSignalSimulator" pkg="hratc2015_framework" type="coilSignalSimulator.py"/>

    <!-- The HRATC judge -->
    <node name="judge" pkg="hratc2015_framework"  type="judge"
      output="screen">
     <param name="~config" value="$(find hratc2015_framework)/src/config/config.ini"/>    
     <param name="~isSimulation" value="true"/>    
    </node>

    <!-- The minefield viewer - updates an occupancy grid displaying the covered area -->
    <node name="mineFieldViewer" pkg="hratc2015_framework"  type="mineFieldViewer"
      output="screen">
     <param name="~config" value="$(find hratc2015_framework)/src/config/config.ini"/>    
     <param name="~isSimulation" value="true"/>    
    </node>

    <!-- Visualization in RVIZ -->
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find hratc2015_framework)/hratc.rviz" /> 
  
</launch>
